crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3277,8 +3278,8 @@ mod tests {
 
         // the type as seen from external modules (`Unknown`)
         // is different from the type inside the module itself (`Unbound`):
-        assert_public_ty(&db, "src/package/foo.py", "x", "Unbound");
-        assert_public_ty(&db, "src/package/bar.py", "x", "Unknown");
+        assert_public_ty(&db, "src/package/foo.py", "x", "Never");
+        assert_public_ty(&db, "src/package/bar.py", "x", "Never");
```

Comment:
I don't think this is a behavior change we want. If you import a fully undeclared/undefined name, you should get a diagnostic, but the imported name should have type `Unknown` in this case. Otherwise you'll just get lots of useless cascading errors about doing invalid things with `Never`.

There is currently a special case in `TypeInferenceBuilder::infer_import_from_definition` where we transform `Unbound` to `Unknown` -- that's why this test exists. That special case becomes obsolete with this PR (we'll never get `Unbound` there anymore), but I think the special case should be updated to check for `Never`, emit a diagnostic (use `self.add_diagnostic`), and transform the `Never` to `Unknown`. (It's also fine if you want to just add a TODO for the diagnostic and we can add it in a later PR; we'll also want a similar diagnostic that's currently missing in `infer_import_definition`, for a nonexistent module.)

Reply:
Just to double-check, `x` as viewed from `bar.py` should be `Unknown` still, not `Never`, right?

Reply:
Right, I think we should handle that in the importing code, instead of the current code which replaces unbound with unknown we'll want to replace never with unknown (and emit a diagnostic, but that can be a TODO if you want, doesn't have to be in this PR.)